discovery-client/src/main/java/io/micronaut/discovery/vault/config/client/v1/VaultConfigConfigurationClientV1.java

```diff
@@ -0,0 +1,189 @@
+
+        final Function<Throwable, Publisher<? extends VaultResponseV1>> errorHandler = getErrorHandler(sourceCount);
+        final List<PairVaultResponse> configurationValuesList = retrieveVaultProperties(activeNames);
+
+        return Flowable.fromIterable(configurationValuesList).concatMapEager(pairVaultResponse -> {
+            return pairVaultResponse.getRight().onErrorResumeNext(errorHandler)
+                    .flatMap(vaultResponse -> Flowable.create(emitter -> {
+
+                        String vaultSourceName = getVaultSourceName(activeNames, pairVaultResponse.getLeft());
+                        Map<String, Object> vaultResponseData = vaultResponse.getData();
+
+                        synchronized (sourceCount) {
+                            sourceCount.getAndIncrement();
+                        }
+
+                        if (!CollectionUtils.isEmpty(vaultResponseData)) {
+                            if (LOG.isInfoEnabled()) {
+                                LOG.info("Obtained property source from Vault, source={}", vaultSourceName);
+                            }
+                            emitter.onNext(PropertySource.of(vaultSourceName,
+                                    vaultResponseData, Integer.MAX_VALUE - activeNamesList.indexOf(
+                                            pairVaultResponse.getLeft())));
+                        }
+
+                        //if all items have been processed, emit onComplete
+                        if (sourceCount.get() == configurationValuesList.size()) {
+                            emitter.onComplete();
+                        }
+            }, BackpressureStrategy.ERROR));
+        });
+    }
+
+    /**
+     * @param activeNames active environment names
+     * @return list of responses from vault
+     */
+    private List<PairVaultResponse> retrieveVaultProperties(Set<String> activeNames) {
+        final String applicationName = getApplicationConfiguration().getName().get();
+
+        return activeNames.stream().map(activeName -> new PairVaultResponse(activeName,
+                    activeName.equals(applicationName) ?
+                        Flowable.fromPublisher(
+                                vaultConfigClientV1.readConfigurationValues(
+                                    getVaultClientConfiguration().getBackend(), applicationName)) :
+                        Flowable.fromPublisher(
+                                vaultConfigClientV1.readConfigurationValues(
+                                    getVaultClientConfiguration().getBackend(), applicationName, activeName))
+                )).collect(Collectors.toList());
+    }
+
+    /**
+     * @param sourceCount total of property sources from vault
+     * @return the error handler to handle vault failed requests
+     */
+    private Function<Throwable, Publisher<? extends VaultResponseV1>> getErrorHandler(AtomicInteger sourceCount) {
+        return throwable -> {
+
+            synchronized (sourceCount) {
```

Comment:
This synchronization block is unnecessary

Reply:
This was a trick one, the only way to now that all sources have been processed I had to include a counter and then use this counter to emit onComplete. I will try without the synchronized and see if it works, during my tests it didn't.

Reply:
It's an atomic integer so it is thread safe. There should be a better way to know if all sources are complete than using a counter

Reply:
will work on it